### PR TITLE
[Table] Frozen Rows/Cols - Part 3: Tests for new code

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests

#### Changes proposed in this pull request:

- Add validation for `numFrozenRows` bounds and `numFrozenColumns` bounds.
- Move an existing `Table` warning message into `errors.ts`.
- Fix bug in `Table#syncRowHeaderSize` when `rowHeaderElement == null`.
- Add tests for all the new quadrant sync'ing code.
    - Size sync'ing (e.g. when toggling `isRowHeaderShown` and `useInteractionBar`)
    - Scroll sync'ing

#### Reviewers should focus on:

- Do these tests look okay?
- Is the validation logic for `numFrozenColumns` and `numFrozenRows` okay? Feels slightly excessive but still helpful.
- I couldn't figure out how to write a few of the tests, so I left them as placeholders. LMK if you have ideas. (See: https://github.com/palantir/blueprint/pull/1361/files#diff-eea83e2366941709af74b6efcc2454bfR287)

### CODE REVIEW

File: packages/table/src/table.tsx [hunk 1]
```diff
@@ -666,20 +667,28 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
+                    console.warn(Errors.TABLE_NON_COLUMN_CHILDREN);
                 }
             }
         });
+
+        if (numFrozenColumns != null && (numFrozenColumns < 0 || numFrozenColumns > React.Children.count(children))) {
+            throw new Error(Errors.TABLE_NUM_FROZEN_COLUMNS_BOUND);
+        }
+        if (numFrozenRows != null && (numFrozenRows < 0 || numFrozenRows > props.numRows)) {
+            throw new Error(Errors.TABLE_NUM_FROZEN_ROWS_BOUND);
```
Comment (giladgray) [inline packages/table/src/table.tsx]: i wonder if this error is worthwhile or if we should just do the sane thing of clamping props values to these bounds when using them?

Comment (giladgray) [inline packages/table/src/table.tsx]: like, is it worth complaining to the user when a reasonable alternative path is readily available?

in #981 i replaced most of our validation errors with warnings, or removed them altogether in cases where the component could easily handle an 'invalid' value

Comment (cmslewis) [inline packages/table/src/table.tsx]: I was wondering about this; I'd be happy to just clamp the out-of-bounds values and print a warning.

**Comment by cmslewis** [comment]
@giladgray take a look at the latest commit. I ended up putting `numFrozenColumnsClamped` and `numFrozenRowsClamped` into `this.state`. LMK what you think.

File: packages/table/src/table.tsx [hunk 2]
```diff
@@ -1169,26 +1206,19 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
...
-        showFrozenRowsOnly: boolean = false,
-        showFrozenColumnsOnly: boolean = false,
-        // rowIndexStart?: number,
-        // rowIndexEnd?: number,
-        // columnIndexStart?: number,
-        // columnIndexEnd?: number,
-        // numFrozenColumns?: number,
-        // numFrozenRows?: number,
-    ) => {
+    private renderBody = (showFrozenRowsOnly: boolean = false, showFrozenColumnsOnly: boolean = false) => {
```
Comment (cmslewis) [inline packages/table/src/table.tsx]: This was just an unrelated cleanup.

File: packages/table/src/table.tsx [hunk 3]
```diff
@@ -506,6 +529,8 @@ export class Table extends AbstractComponent<ITableProps, ITableState> {
         this.setState({
             columnWidths: newColumnWidths,
             focusedCell: enableFocus ? newFocusedCellCoordinates : undefined,
+            numFrozenColumnsClamped: this.getNumFrozenColumnsClamped(nextProps),
```
Comment (gscshoyru) [inline packages/table/src/table.tsx]: I feel a little weird storing stuff in state that can be otherwise be derived from state/props. This is the whole "multiple sources of truth" problem. I'd much rather have a private function that just returns this number. I can be convinced otherwise -- but it's probably better to do that way.

Comment (cmslewis) [inline packages/table/src/table.tsx]: I like the idea of a private function too. But really I'd prefer to just get it from `this.props`, hence my approach of throwing an error before. :/
